keps/sig-auth/2799-token-controller-deprecation/README.md

```diff
@@ -0,0 +1,477 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a namespace-scoped and service-account-scoped binary label to
+allow users to opt in/out the provision of secret-based service account tokens
+in Token Controller. In addition, it sketches out the phases to deprecate Token
+Controller.
+
+## Motivation
+
+As BoundServiceAccountTokenVolume is GA in 1.22, pods’ service account tokens
+would be obtained via TokenRequest API and stored as projected volume. This
+change obviates the need for Token Controller’s reconciling of secret-based
+service account tokens. The secret-based tokens are [not secure by design](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#background)
+and the token controller is [fragile by design in some cases](https://github.com/kubernetes/kubernetes/issues/98474)
+where it is unable to handle the churns between secrets and service account
+controller loops.
+
+### Goals
+
+- Provides a knob to opt in/out the generation of secret-based token for a
+  service account
+- Provides a knob to opt in/out the generation of secret-based token for all
+  service accounts in a namespace
+- Deprecate Token Controller
+- No Forever Token in Kubernetes
+
+### Non-Goals
+
+## Proposal
+
+With the goal to deprecate Token Controller,
+
+- Phase I: At release N, `serviceaccount.kubernetes.io/reconcile-legacy-token` 
+  is `true` by default to align with existing behavior of Token Controller, i.e.
+  Token Controller will continue to generate tokens as it is now. To disable the
+  generation of secret-based tokens, users are required to specify
+  `serviceaccount.kubernetes.io/reconcile-legacy-token` for the service accounts
+  or namespaces that they would like to prevent the generation of secret-based
+  tokens. To spread the “fire”, in-tree service accounts in `kube-system` will 
+  be annotated as `false`. At the same time, warn usage of secret-based tokens.
+  This phase creates leeway for users to migrate off their use of secret-based
+  service account tokens.
+- Phase II: At release N+X (X>0), `serviceaccount.kubernetes.io/reconcile-legacy-token`
+  is `false` by default. Token Controller will no longer generate secret-based token
+  unless `serviceaccount.kubernetes.io/reconcile-legacy-token`. In addition, 
+  keep warn usage of legacy tokens that in future releases, secret-based tokens
+  would be purged.
+- Phase III: At release N+Y (Y>X), begin to purge all secret-based tokens in existing
+  clusters unless the referenced service accounts or namespaces have
+  `serviceaccount.kubernetes.io/reconcile-legacy-token`.
+- Phase IV: At release N+Z (Z>Y), remove the annotation and all
+  secret-based tokens are purged.
+- Phase V: At release N+A (A>Z), Token Controller and secret-based token authenticator
```

Comment:
There are two parts of service account tokens.  Implicit tokens (generated for you), which we are changing.  Explicit tokens (user request token for service account on a secret) are part of v1 workload api and i don’t think we can safely remove, unless you mean “repurpose token controller”.  Or are you proposing building a bound token into that secret?

Reply:
for the user requested tokens, we probably want to start encouraging users to use TokenRequest API.

Reply:
https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-a-service-account-api-token

Reply:
+1 to not breaking explicit tokens.